nex/token/reward.py

```diff
@@ -1,97 +1,123 @@
+    """
+    Loop up the level of the owner of address. Then raise it
+    by 1, put it back into storage and return the new value.
     """
     
-    return True
+    storage = StorageAPI()
+    
+    level_key = storage.get_level_key(address)
+    
+    level = level_of(level_key)
+    
+    new_level = level + 1
+    
+    storage.put(level_key, new_level)
+    
+    return new_level
 
 
 def calculate_reward(address):
     """
+    Apply the calculation function for the user reward.
     """
     
-    ## make use of 'level_of' below
+    storage = StorageAPI()
     
-    """
-    C# REFERENCE IMPLEMENTATION                 <====== WE DO IT BETTER THIS TIME, THOUGH
+    level_key = storage.get_level_key(address)
     
-            /// Compute user reward depending on user level
-            
-            BigInteger basicReward = 100000000; 
-            // TODO test if I can just use 'factor' here
-            BigInteger bonusFactor = 1;
-
-            /// Quick level up for demonstration purpose before official launch
-            if (level > 1) bonusFactor = 2;
-            if (level > 3) bonusFactor = 5;
-            if (level > 9) bonusFactor = 20;
-
-            return bonusFactor * basicReward;
-    """
+    level = level_of(address)
     
-    return True
+    bonus_factor = 1
     
+    if level > 1: 
+        bonus_factor = 2
+    if level > 3 
+        bonus_factor = 5
+    if level > 9:
+        bonus_factor = 20
+    
+    basic_reward = 100000000
+    
+    reward = basic_reward * bonus_factor
+        
+    return reward
 
-def level_up(address):
+
+def reward_user(address):
     """
+    Raise the user level corresponding to address and reward him/her.
     """
     
-    ## make use of 'level_up' below
+    token = Token()
     
-    """
-    C# REFERENCE IMPLEMENTATION
+    if not CheckWitness(token.owner):
+        print("Must be owner to reward")
+        return False
     
-            BigInteger newLevel = LevelOf(account) + 1;
-            
-            Storage.Put(Storage.CurrentContext, Key("L", account), newLevel);
-            
-            return newLevel;
-    """
+    nep = NEP5Handler()
+        
+    level = level_up(address) ### you might wanna just set this constant for debugging purposes
+        
+    reward = calculate_reward(level)
     
-    return True
+    success = nep.do_transfer(storage, token.owner, address, reward)
     
+    return success
 
-def level_of(address):
+
+def reward_users(addresses, reward):
     """
+    Raise the user level corresponding to address and reward him/her.
     """
     
-    ## to be called from Main upon 'operation == "level"' if called by master wallet
+    token = Token()
     
-    """
-    C# REFERENCE IMPLEMENTATION
-    
-            BigInteger nLevel = 0;
-
-            byte[] level = Storage.Get(Storage.CurrentContext, Key("L", account));
-
-            if (level != null)
-                nLevel = level.AsBigInteger();
+    if not CheckWitness(token.owner):
```

Comment:
While I get this for the general implementation, I don't think we need to check the owner. Anyone who wants to send funds to all users as a reward is welcome to! ;-)